### PR TITLE
Fix false negative for `used-before-assignment` when some except handlers don't define a name

### LINKED ISSUE #5524: False-negative `used-before-assignment` after try-except block
Followup to https://github.com/PyCQA/pylint/issues/5500#issuecomment-991695275

```py
def func(var):
    try:
        return 1 / var.some_other_func()
    except AttributeError:
        pass
    except ZeroDivisionError:
        msg = "Devision by 0"

    print(msg)  # should emit 'used-before-assignment'
```

The code above should raise a `used-before-assignment` error since `msg` is not assigned in the `AttributeError` ExceptHandler.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Fixes #5524.

<s>In discussion I suggested closing #2835 as a duplicate, but if folks like we could "unlink" the two and reopen the older ticket.

The older ticket is just suggesting "always assume try statements fail", and @cdce8p suggested a new message for that (`possibly-unbound`). Probably a 2.14 or later thing. By contrast this PR is just fixing a quick false negative. Sorry for not seeing that we were more likely to tackle these separately!</s> Edit: done!

### CODE REVIEW

File: pylint/checkers/variables.py
Comment (areveny) [inline pylint/checkers/variables.py:736]: Hey, I see you're still adding commits but some of your test failures might be from this `.parent.parent.handlers` pattern, which might not be resilient to cases where there are parent statements between the name and the `try`/`except`.

It looks like some failures are coming from [ntpath](https://github.com/python/cpython/blob/main/Lib/ntpath.py) line 467 where an entire function is defined under an `except`.

Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:736]: Hey, thanks for having a look. I _should_ be okay with the two-parent pattern, since it was just checked earlier in this function. But at the same time I share your suspicion that something could be up with that assumption -- maybe we might be missing some more false negatives?

Comment (areveny) [reply pylint/checkers/variables.py:736]:
```
def func_invalid3(var):
    """'msg' is not defined in one handler."""
    err_message = False
    try:
        return 1 / var.some_other_func()
    except AttributeError:
        msg: str
    except ZeroDivisionError:
        if err_message:
            msg = "Division by 0"
        else:
            msg = None
    print(msg)  # [used-before-assignment]
```

Yeah, this would be a false negative with the check.

Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:736]: You're right, this is another "don't assume direct parentage" false negative like #5582. I can address it on this branch.

Comment (jacobtylerwalls) [reply pylint/checkers/variables.py]: Test case added in 054fb68

File: pylint/checkers/variables.py
Comment (areveny) [inline pylint/checkers/variables.py:754]:
```suggestion
            if isinstance(stmt, nodes.Assign):
                 for target in stmt.targets:
                     if isinstance(target, nodes.AssignName) and target.name == name:
                         return True
                     if isinstance(target, (nodes.Tuple, nodes.List)):
                         for elt in utils.get_all_elements(target):
                             if isinstance(elt, nodes.Const) and elt.value == name:
                                 return True
```

Three parts to this suggestion. 

1) I figured out the case where `len(stmt.targets) > 1`. Looping over `stmt.targets` catches this case:

```
except ZeroDivisionError:
   bsg = msg = "Division by 0"
print(msg)
``` 

2) `as_string` is primarily used to format message parameters, and it seems better to avoid it in comparisons. 

3) Check target unpacking to `List` and `Tuple`, which can recursively contain more `List` or `Tuple` to catch this case:

```
asd, msg = ("asd", "Division by 0")
[asd, msg] = ["asd", "Division by 0"]
```

File: pylint/checkers/variables.py
Comment (areveny) [inline pylint/checkers/variables.py:763]:
```suggestion
                    for arg_or_kwarg in stmt.test.args + [keyword.value for key word in stmt.test.keywords]:
```

`keywords` is an array of `nodes.Keyword` wrapping `nodes.NamedExpr`.

```
    except AttributeError:
        if print(bsg := var.get_msg(), sep=(msg:=var.get_bsg())):
            pass
```

File: pylint/checkers/variables.py
Comment (jacobtylerwalls) [reply pylint/checkers/variables.py]: Thank you, this is fantastic help.

**Review by areveny — APPROVED** [review]
Great! Thanks for the changes and really nice simplification with the last commit.

I'm satisfied with this, if we could get a second approval I think this is ready to merge. Thanks!

File: tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py
Comment (DanielNoord) [inline tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: These tests are great! Just one remark: `func_invalid3` is not actually testing anything new compared to `func_invalid2`. I think we want to test the assignment in `if` statements here, which would be best tested by removing the `AttributeError` except handler and checking if not message is being raised. Currently this just tests that `msg: str` works as expected, if the handling of the `if` statement is incorrect we wouldn't notice.

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:706]: Gotta love this `util` 😄

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:714]: I think it would be better to add an ignore wherever that is necessary and then wait for `astroid`'s typing to be updated. If we update `parent` of `ExceptHandler` at some point we wouldn't notice that this typing is redundant.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Small changes, but this looks great!

@jacobtylerwalls just to get this right, what do you propose in terms of issue management. What should be opened/closed after this has been merged?

File: tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py
Comment (jacobtylerwalls) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: Actually, I think it does test something different. The first implementation didn't catch this case (by @areveny), and we discussed it [here](https://github.com/PyCQA/pylint/pull/5764#discussion_r798116392).

Should I flesh out the docstring to clarify?

File: pylint/checkers/variables.py
Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:714]: `mypy` didn't make me do this, I just thought I was helping readability! I'll change this to an assert instead to make more explicit that I'm depending on the parent being a TryExcept.

Comment (DanielNoord) [reply pylint/checkers/variables.py]: Ah okay, never mind then. I'm not sure about performance impact but typing is probably better than asserts? So we should probably revert 😅

File: tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py
Comment (DanielNoord) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: Hmm could it be that something changed in the implementation that made that test different? 

```python
def func_invalid3(var):
    """'msg' is not defined in one handler."""
    err_message = False
    try:
        return 1 / var.some_other_func()
    except ZeroDivisionError:
        if err_message:
            msg = "Division by 0"
        else:
            msg = None
    print(msg)
```

This shouldn't raise right?

Comment (jacobtylerwalls) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: Sorry, I don't think I'm following you very well. 😅 . I thought you were proposing I take `func_invalid3` and rewrite it like you just did in the comment directly above this one. But I'm saying that `func_invalid3` as written catches a false negative that this PR didn't address until @areveny pointed it out and I fixed it (to stop assuming direct parentage), so I think the test has value as written. But I think I may be missing your point still, perhaps? Would you like to me _add_ this as another test, so it becomes `func_ok6`?

Comment (DanielNoord) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: Yeah I'm not making many sense. Let me rephrase:
I don't really understand why the third function needs to be tested in combination with the `msg: str` statement. Why would the behaviour of the message change when there is a combination of `msg: str` and `ifs` in two expect handlers. Imo it makes more sense to test those separately.
However, you have discussed this a lot with @areveny so perhaps there is something difficult going on when both of those patterns are present and we should test them together? Then perhaps the docstring should be added.
While trying to find out if I could find this relatedness between `msg: str` and the `ifs` I stumbled upon the example I gave above. I think the current version raises on this code but it shouldn't, right? That is a separate issue from the one with which I started this comment thread, but it might be related.

Sorry for being so unclear 😅 Long day..

Comment (jacobtylerwalls) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: No problem, I can explain.

> However, you have discussed this a lot with @areveny so perhaps there is something difficult going on when both of those patterns are present and we should test them together?

In a word, yup.

I can change `msg: str` to `pass` if that would be more clear that it doesn't have to do with type annotations. However, having two handlers where the one with the name is nested under if or with is the essence of the test case. All of the tests for correct positives require multiple except handlers. This one in involves nesting under an if or with just to catch the faulty assumption the code first made about direct parentage. 

"Round 1" of most of my control flow PRs adding methods like `_uncertain_nodes_in...` just assumed direct parentage of the node being assigned by its containing try/except. After you pointed me to `utils.get_node_first_ancestor_of_type` I did #5582 to stop assuming this and catch more false negatives, but that PR didn't touch this method. (I guess I was trying to take small bites! 😅 ) @areveny spotted the false negative, and I decided it should be done in this PR. So if you check out the first commit on the branch, you'll see it fails the test case as written.

***
And then you found a false positive for me to investigate! I'll see what's up with that.

Comment (DanielNoord) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:100]: > I can change `msg: str` to `pass` if that would be more clear that it doesn't have to do with type annotations. However, having two handlers where the one with the name is nested under if or with is the essence of the test case. All of the tests for correct positives require multiple except handlers. This one in involves nesting under an if or with just to catch the faulty assumption the code first made about direct parentage.

But if there was only a single expect handler with nested if and the message isn't raised, won't the functionality of nested ifs be tested as well? If it does not work (ie., assignment in nested ifs is not recognised) the message would be raised incorrectly. 

> And then you found a false positive for me to investigate! I'll see what's up with that.

👍

Comment (jacobtylerwalls) [reply tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py:124]: > But if there was only a single expect handler with nested if and the message isn't raised, won't the functionality of nested ifs be tested as well? If it does not work (ie., assignment in nested ifs is not recognised) the message would be raised incorrectly.

Ah, no, I don't think testing a correct positive and correct negative will necessarily go down the same code path. You're talking about testing a correct negative (which is good to test!), but that will short-circuit a lot earlier (see `"# If the other node is in the same scope as this node, assume it executes"`), which is why it was probably passing.

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:752]: Should we add typing for `stmt` here?

**Review by DanielNoord — COMMENTED** [review]
One final question!

**Comment by jacobtylerwalls** [comment]
@areveny @DanielNoord thanks for the reviews and bugspotting. I'm glad we're getting a real v1.0 of control flow for try/except into something that should _just work_!

> in terms of issue management

Let's pretend I never called the two issues duplicates! Specifically:
- reopen #2835 
- paste a link there to Marc's "possibly unbound" suggestion in https://github.com/PyCQA/pylint/issues/5524#issuecomment-1025262104
- I'll leave a comment there that the implementation can reuse `NamesConsumer._defines_name_raises_or_returns`

And then I can adjust this PR to say "closes" instead of "partially closes" #5524. Sorry for the extra work, but now I see from discussion we're likely to address the bigger-picture issue with a new message, so better to keep it all separate.

**Comment by DanielNoord** [comment]
I have opened and made the comments as you suggested. I'm going to ask for a review from Pierre as this is quite a large PR again.

Just FYI: I think this should be the last new PR to go into `2.13`. We really (really really really) need to release as we're hitting almost 300 issues/PRs closed with a lot of useful features as well. The release is also blocking some important maintenance work such as the `argparse` refactor.
Any help with any of the open `2.13` issues would be much appreciated. Especially with #5645 as I don't have the necessary hardware to work on it myself...

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — COMMENTED** [review]
LGTM, the tests are really nice, and this will be a nice addition to 2.13. Also thank you for providing early feedback @areveny, :)

Small question, I noticed that there could be more tests if we do the permutation for "defined in one handler only" with unpacked iterable, or chained assignment. I suppose it's because it's not necessary to check this again as it's already checked in funcok_4 and func_ok5, and not doing all permutation would speed the functional tests up ?

**Comment by jacobtylerwalls** [comment]
I went ahead and added the tests. After all, just a few comments above I was very keen to tell the world that true positives and true negatives might hit different code paths! 😄

**Pull request merged**
